tikv/write.go

```diff
@@ -214,114 +177,212 @@ func (w *writeLockWorker) run() {
+		writer:  writer,
+	}.run()
+
+	go rollbackGCWorker{
+		writer: writer,
+	}.run()
+}
+
+func (writer *dbWriter) Close() {
+	close(writer.closeCh)
+	writer.wg.Wait()
+}
+
+func (writer *dbWriter) Write(batch mvcc.WriteBatch) error {
+	wb := batch.(*writeBatch)
+	if len(wb.dbBatch.entries) > 0 {
+		// We must delete lock after commit succeed, or there will be inconsistency.
+		wb.dbBatch.wg.Add(1)
+		writer.dbCh <- &wb.dbBatch
+		wb.dbBatch.wg.Wait()
+		err := wb.dbBatch.err
+		if err != nil {
+			return err
+		}
 	}
-	err = f.Sync()
-	if err != nil {
-		return errors.Trace(err)
+	if len(wb.lockBatch.entries) > 0 {
+		wb.lockBatch.wg.Add(1)
+		writer.lockCh <- &wb.lockBatch
+		wb.lockBatch.wg.Wait()
+		return wb.lockBatch.err
 	}
-	f.Close()
-	return os.Rename(tmpFileName, store.dir+"/lock_store")
+	return nil
+}
+
+type writeBatch struct {
+	startTS   uint64
+	commitTS  uint64
+	dbBatch   writeDBBatch
+	lockBatch writeLockBatch
+}
+
+func (wb *writeBatch) Prewrite(key []byte, lock *mvcc.MvccLock) {
+	wb.lockBatch.set(key, lock.MarshalBinary())
 }
 
-func (store *MVCCStore) loadLocks() error {
-	fileName := store.dir + "/lock_store"
-	f, err := os.Open(fileName)
-	if err != nil {
-		if os.IsNotExist(err) {
-			return nil
+func (wb *writeBatch) Commit(key []byte, lock *mvcc.MvccLock) {
+	if lock.Op != uint8(kvrpcpb.Op_Lock) {
+		userMeta := mvcc.NewDBUserMeta(wb.startTS, wb.commitTS)
+		wb.dbBatch.set(key, lock.Value, userMeta)
+		if lock.HasOldVer {
+			oldKey := mvcc.EncodeOldKey(key, lock.OldMeta.CommitTS())
+			wb.dbBatch.set(oldKey, lock.OldVal, lock.OldMeta.ToOldUserMeta(wb.commitTS))
 		}
-		return errors.Trace(err)
 	}
-	defer f.Close()
-	reader := bufio.NewReader(f)
-	hdrBuf := make([]byte, 8)
-	hdr := (*lockEntryHdr)(unsafe.Pointer(&hdrBuf[0]))
-	var keyBuf, valBuf []byte
-	for {
-		_, err = reader.Read(hdrBuf)
-		if err == io.EOF {
+	wb.lockBatch.delete(key)
+}
+
+func (wb *writeBatch) Rollback(key []byte, deleteLock bool) {
+	rollbackKey := mvcc.EncodeRollbackKey(nil, key, wb.startTS)
+	wb.lockBatch.rollback(rollbackKey)
+	if deleteLock {
+		wb.lockBatch.delete(key)
+	}
+}
+
+func (writer *dbWriter) NewWriteBatch(startTS, commitTS uint64) mvcc.WriteBatch {
+	if commitTS > 0 {
+		writer.updateLatestTS(commitTS)
+	} else {
+		writer.updateLatestTS(startTS)
+	}
+	return &writeBatch{
+		startTS:  startTS,
+		commitTS: commitTS,
+	}
+}
+
+func (writer *dbWriter) getLatestTS() uint64 {
+	return atomic.LoadUint64(&writer.latestTS)
+}
+
+func (writer *dbWriter) updateLatestTS(ts uint64) {
+	latestTS := writer.getLatestTS()
+	if ts != math.MaxUint64 && ts > latestTS {
+		atomic.CompareAndSwapUint64(&writer.latestTS, latestTS, ts)
+	}
+}
+
+const delRangeBatchSize = 4096
+
+func (writer *dbWriter) DeleteRange(startKey, endKey []byte, latchHandle mvcc.LatchHandle) error {
+	keys := make([][]byte, 0, delRangeBatchSize)
+	oldStartKey := mvcc.EncodeOldKey(startKey, maxSystemTS)
+	oldEndKey := mvcc.EncodeOldKey(endKey, maxSystemTS)
```

Comment:
If the range is [start, end), i think you don't need to calculate oldEndKey, just use endkey for old cf.

Reply:
But `endKey` has prefix 't', `oldStartKey` and `oldEndKey` has prefix `u`.